File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/Anthropic.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.Anthropic;

/// <summary>Structural client interface — one method per Anthropic operation.</summary>
public interface IAnthropicClient
{
    AnthropicResponse MessagesCreate(JsonObject @params);
}

/// <summary>Anthropic response — minimal subset of fields we need.</summary>
public sealed class AnthropicResponse
{
    public string Model { get; }
    public long InputTokens { get; }
    public long OutputTokens { get; }
    public IReadOnlyList<ContentBlock> Content { get; }

    public AnthropicResponse(string model, long inputTokens, long outputTokens,
                             IReadOnlyList<ContentBlock>? content)
    {
        Model = model;
        InputTokens = inputTokens;
        OutputTokens = outputTokens;
        Content = content ?? new List<ContentBlock>();
    }
}

/// <summary>A content block in a response.</summary>
public sealed class ContentBlock
{
    public string Type { get; }
    public ContentBlock(string type) { Type = type; }
}

/// <summary>Options for <see cref="BoundedAnthropicClient"/>.</summary>
public sealed class AnthropicOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static AnthropicOptions Defaults() => new();
}

/// <summary>
/// Wraps an <see cref="IAnthropicClient"/> so every call is accounted
/// against a <see cref="BoundedExecution"/>.
/// </summary>
public sealed class BoundedAnthropicClient
{
    private readonly IAnthropicClient _raw;
    private readonly BoundedExecution _execution;
    private readonly AnthropicOptions _opts;

    public BoundedAnthropicClient(IAnthropicClient raw, BoundedExecution execution, AnthropicOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? AnthropicOptions.Defaults();
    }

    public AnthropicResponse MessagesCreate(JsonObject @params)
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();

        AnthropicResponse resp;
        try { resp = _raw.MessagesCreate(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        long tokens = resp.InputTokens + resp.OutputTokens;
        long toolCalls = 0;
        foreach (var b in resp.Content) if (b.Type == "tool_use") toolCalls++;
        string model = string.IsNullOrEmpty(resp.Model) ? "unknown" : resp.Model;
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"anthropic.messages.create:{model}",
        });
        return resp;
    }
}

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoMiddleware/Anthropic.swift
import Foundation
import IAIsoCore

/// Anthropic Messages API middleware.
public enum Anthropic {

    /// Structural client interface — one method per Anthropic operation.
    public protocol Client {
        func messagesCreate(_ params: [String: Any]) throws -> Response
    }

    public struct ContentBlock: Sendable, Equatable {
        public let type: String
        public init(type: String) { self.type = type }
    }

    public struct Response: Sendable {
        public let model: String
        public let inputTokens: Int64
        public let outputTokens: Int64
        public let content: [ContentBlock]

        public init(
            model: String,
            inputTokens: Int64,
            outputTokens: Int64,
            content: [ContentBlock]
        ) {
            self.model = model
            self.inputTokens = inputTokens
            self.outputTokens = outputTokens
            self.content = content
        }
    }

    public struct Options: Sendable {
        public var raiseOnEscalation: Bool
        public init(raiseOnEscalation: Bool = false) {
            self.raiseOnEscalation = raiseOnEscalation
        }
        public static let defaults = Options()
    }

    /// Wraps a `Client` so every call is accounted against a `BoundedExecution`.
    public final class BoundedClient {
        private let raw: Client
        private let execution: BoundedExecution
        private let opts: Options

        public init(raw: Client, execution: BoundedExecution, options: Options = .defaults) {
            self.raw = raw
            self.execution = execution
            self.opts = options
        }

        public func messagesCreate(_ params: [String: Any]) throws -> Response {
            let pre = execution.check()
            if pre == .locked { throw MiddlewareError.locked }
            if pre == .escalated, opts.raiseOnEscalation {
                throw MiddlewareError.escalationRaised
            }
            let resp: Response
            do {
                resp = try raw.messagesCreate(params)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            let tokens = resp.inputTokens + resp.outputTokens
            let toolCalls: Int64 = resp.content.reduce(0) { acc, b in
                acc + (b.type == "tool_use" ? 1 : 0)
            }
            let model = resp.model.isEmpty ? "unknown" : resp.model
            execution.recordStep(StepInput(
                tokens: tokens,
                toolCalls: toolCalls,
                tag: "anthropic.messages.create:\(model)"))
            return resp
        }
    }
}
